File: storio-sqlite-annotations-processor/src/main/java/com/pushtorefresh/storio/sqlite/annotations/processor/generate/TableGenerator.java
package com.pushtorefresh.storio.sqlite.annotations.processor.generate;

import com.pushtorefresh.storio.common.annotations.processor.generate.Generator;
import com.pushtorefresh.storio.sqlite.annotations.processor.introspection.StorIOSQLiteColumnMeta;
import com.pushtorefresh.storio.sqlite.annotations.processor.introspection.StorIOSQLiteTypeMeta;
import com.squareup.javapoet.ClassName;
import com.squareup.javapoet.FieldSpec;
import com.squareup.javapoet.JavaFile;
import com.squareup.javapoet.MethodSpec;
import com.squareup.javapoet.ParameterSpec;
import com.squareup.javapoet.TypeName;
import com.squareup.javapoet.TypeSpec;

import org.jetbrains.annotations.NotNull;

import java.util.ArrayList;
import java.util.Collection;
import java.util.Collections;
import java.util.Comparator;
import java.util.List;

import javax.lang.model.element.AnnotationMirror;
import javax.lang.model.element.Modifier;

import static com.pushtorefresh.storio.common.annotations.processor.generate.Common.INDENT;

public class TableGenerator implements Generator<StorIOSQLiteTypeMeta> {

    private static final String SUFFIX = "Table";
    private static final String CREATE_METHOD_NAME = "createTable";
    private static final String UPDATE_METHOD_NAME = "updateTable";
    private static final String UPDATE_METHOD_VERSION_OLD_PARAM = "versionOld";
    private static final String METHOD_DB_PARAM = "db";

    private final ClassName db = ClassName.get("android.database.sqlite", "SQLiteDatabase");
    private static final String tableName = "tableName";

    @Override
    public JavaFile generateJavaFile(@NotNull StorIOSQLiteTypeMeta typeMeta) {
        if (!typeMeta.storIOType.generateTableClass()) {
            return null;
        }

        TypeSpec tableSpec = TypeSpec.classBuilder(typeMeta.simpleName + SUFFIX)
                .addFields(generateFields(typeMeta.storIOType.table(), typeMeta.columns.values()))
                .addMethod(generateCreateMethod(typeMeta.storIOType.table(), typeMeta.columns.values()))
                .addMethod(generateUpdateMethod(typeMeta.storIOType.table(), typeMeta.columns.values()))
                .addModifiers(Modifier.PUBLIC, Modifier.FINAL)
                .addMethod(generateConstructor())
                .build();
        return JavaFile.builder(typeMeta.packageName, tableSpec)
                .indent(INDENT)
                .build();
    }

    @NotNull
    private Iterable<FieldSpec> generateFields(@NotNull String table, @NotNull Collection<StorIOSQLiteColumnMeta> columns) {
        List<FieldSpec> list = new ArrayList<FieldSpec>();
        list.add(FieldSpec.builder(String.class, tableName)
                .initializer("$S", table)
                .addModifiers(Modifier.PUBLIC, Modifier.STATIC, Modifier.FINAL)
                .build());
        for (StorIOSQLiteColumnMeta column : columns) {
            list.add(FieldSpec.builder(String.class, "COLUMN_" + column.elementName)
                    .initializer("$S", column.storIOColumn.name())
                    .addModifiers(Modifier.PUBLIC, Modifier.STATIC, Modifier.FINAL)
                    .build());
        }
        return list;
    }

    @NotNull
    private MethodSpec generateConstructor() {
        return MethodSpec.constructorBuilder().addModifiers(Modifier.PRIVATE).build();
    }

    @NotNull
    private MethodSpec generateCreateMethod(@NotNull String table, @NotNull Collection<StorIOSQLiteColumnMeta> columns) {
        String code = "CREATE TABLE " + table + " (";
        int i = 0;
        for (StorIOSQLiteColumnMeta entry : columns) {
            code += entry.storIOColumn.name() + " " + entry.javaType.getSqlType();
            if (isNotNull(entry)) {
                code += " NOT NULL";
            }
            if (entry.storIOColumn.key()) {
                code += " PRIMARY KEY";
            }
            if (i != columns.size() - 1) {
                code += ",\n";
            }
            i++;
        }
        code += ");";
        return MethodSpec.methodBuilder(CREATE_METHOD_NAME)
                .addModifiers(Modifier.PUBLIC, Modifier.STATIC)
                .addParameter(ParameterSpec.builder(db, METHOD_DB_PARAM).build())
                .addStatement(METHOD_DB_PARAM + ".execSQL($S)", code)
                .build();
    }

    @NotNull
    private MethodSpec generateUpdateMethod(@NotNull String table, @NotNull Collection<StorIOSQLiteColumnMeta> columns) {
        List<StorIOSQLiteColumnMeta> list = new ArrayList<StorIOSQLiteColumnMeta>(columns);
        //sort on version
        Collections.sort(list, new Comparator<StorIOSQLiteColumnMeta>() {
            @Override
            public int compare(StorIOSQLiteColumnMeta t0, StorIOSQLiteColumnMeta t1) {
                return t0.storIOColumn.version() - t1.storIOColumn.version();
            }
        });
        //remove version == 0, those are create during createTable
        List<StorIOSQLiteColumnMeta> filtered = new ArrayList<StorIOSQLiteColumnMeta>();
        for (StorIOSQLiteColumnMeta o : list) {
            if (o.storIOColumn.version() != 0) {
                filtered.add(o);
            }
        }
        String code = "";
        if (filtered.size() > 0) {
            for (int i = 0; i < filtered.size(); i++) {
                StorIOSQLiteColumnMeta o = filtered.get(i);
                if (o.storIOColumn.version() > 1) {
                    code += "if(" + UPDATE_METHOD_VERSION_OLD_PARAM + " < " + o.storIOColumn.version() + ") {\n" +
                            "    " + METHOD_DB_PARAM + ".execSQL(\"ALTER TABLE " + table + " ADD COLUMN " + o.storIOColumn.name() + " " + o.javaType.getSqlType();
                    if (isNotNull(o)) {
                        code += " NOT NULL";
                    }
                    code += ";\");\n";
                    code += "}\n";
                }
            }

        }
        return MethodSpec.methodBuilder(UPDATE_METHOD_NAME)
                .addModifiers(Modifier.PUBLIC, Modifier.STATIC)
                .addParameter(ParameterSpec.builder(db, METHOD_DB_PARAM).build())
                .addParameter(ParameterSpec.builder(TypeName.INT, UPDATE_METHOD_VERSION_OLD_PARAM).build())
                .addCode(code)
                .build();
    }

    private boolean isNotNull(@NotNull StorIOSQLiteColumnMeta entry) {
        for (AnnotationMirror mirror : entry.element.getAnnotationMirrors()) {
            // android.support.annotation class is not in the classpath of annotation processor so I check it by name
            if (mirror.getAnnotationType().toString().equals("android.support.annotation.NonNull")) {
                return true;
            }
        }
        return entry.element.getAnnotation(NotNull.class) != null;
    }
}

File: storio-sqlite-annotations-processor/src/main/kotlin/com/pushtorefresh/storio3/sqlite/annotations/processor/generate/TableGenerator.kt
package com.pushtorefresh.storio3.sqlite.annotations.processor.generate

import com.pushtorefresh.storio3.common.annotations.processor.generate.Common.INDENT
import com.pushtorefresh.storio3.common.annotations.processor.generate.Generator
import com.pushtorefresh.storio3.common.annotations.processor.toUpperSnakeCase
import com.pushtorefresh.storio3.sqlite.annotations.processor.introspection.StorIOSQLiteColumnMeta
import com.pushtorefresh.storio3.sqlite.annotations.processor.introspection.StorIOSQLiteTypeMeta
import com.squareup.javapoet.*
import javax.lang.model.element.Modifier.*

private const val ANDROID_NON_NULL_ANNOTATION = "android.support.annotation.NonNull"
private const val DB_PARAM = "db"
private const val OLD_VERSION_PARAM = "oldVersion"

object TableGenerator : Generator<StorIOSQLiteTypeMeta> {

    private val sqliteDatabase = ClassName.get("android.database.sqlite", "SQLiteDatabase")

    override fun generateJavaFile(typeMeta: StorIOSQLiteTypeMeta): JavaFile {
        val tableSpec = TypeSpec.classBuilder("${typeMeta.simpleName}Table")
                .addModifiers(PUBLIC, FINAL)
                .addMethod(generatePrivateConstructor())
                .addFields(generateFields(typeMeta.storIOType.table, typeMeta.columns.values))
                .addMethod(generateCreateMethod(typeMeta.storIOType.table, typeMeta.columns.values))
                .addMethod(generateUpdateMethod(typeMeta.storIOType.table, typeMeta.columns.values))
                .build()

        return JavaFile.builder(typeMeta.packageName, tableSpec)
                .indent(INDENT)
                .build()
    }

    private fun generateFields(table: String, columns: Collection<StorIOSQLiteColumnMeta>): Iterable<FieldSpec> {
        val list = mutableListOf<FieldSpec>()

        list += FieldSpec.builder(String::class.java, "NAME")
                .initializer("\$S", table)
                .addModifiers(PUBLIC, STATIC, FINAL)
                .build()

        columns.forEach { column ->
            list += FieldSpec.builder(String::class.java, "${column.elementName.toUpperSnakeCase()}_COLUMN")
                    .initializer("\$S", column.storIOColumn.name)
                    .addModifiers(PUBLIC, STATIC, FINAL)
                    .build()
        }

        return list
    }

    private fun generateCreateMethod(table: String, columns: Collection<StorIOSQLiteColumnMeta>): MethodSpec {
        val builder = StringBuilder()

        builder.append("CREATE TABLE $table (")

        val primaryKeys = columns.filter { it.storIOColumn.key }.toList()

        columns.forEachIndexed { index, column ->
            builder.append("${column.storIOColumn.name} ${column.javaType.sqliteType}")
            if (column.isNotNull()) builder.append(" NOT NULL")
            if (column.storIOColumn.key && primaryKeys.size == 1) builder.append(" PRIMARY KEY")
            if (index != columns.size - 1) builder.append(",\n")
        }

        if (primaryKeys.size > 1) {
            builder.append(",\nPRIMARY KEY(")
            primaryKeys.forEachIndexed { index, key ->
                builder.append(key.storIOColumn.name)
                if (index != primaryKeys.size - 1) builder.append(", ")
            }
            builder.append(")")
        }

        builder.append(");")

        return MethodSpec.methodBuilder("createTable")
                .addModifiers(PUBLIC, STATIC)
                .addParameter(ParameterSpec.builder(sqliteDatabase, DB_PARAM).build())
                .addStatement("$DB_PARAM.execSQL(\$S)", builder.toString())
                .build()
    }

    private fun generateUpdateMethod(table: String, columns: Collection<StorIOSQLiteColumnMeta>): MethodSpec {
        val builder = MethodSpec.methodBuilder("updateTable")
                        .addModifiers(PUBLIC, STATIC)
                        .addParameter(ParameterSpec.builder(sqliteDatabase, DB_PARAM).build())
                        .addParameter(ParameterSpec.builder(TypeName.INT, OLD_VERSION_PARAM).build())

        val columnsToUpdate = columns
                .filter { it.storIOColumn.version > 1 }
                .sortedBy { it.storIOColumn.version }

        columnsToUpdate.forEach { column ->
            builder.beginControlFlow("if ($OLD_VERSION_PARAM < ${column.storIOColumn.version})")
            builder.addCode("$DB_PARAM.execSQL(\$S);\n", "ALTER TABLE $table ADD COLUMN ${column.storIOColumn.name} ${column.javaType.sqliteType}${if (column.isNotNull()) " NOT NULL" else ""}")
            builder.endControlFlow()
        }

        return builder.build()
    }

    private fun generatePrivateConstructor() = MethodSpec.constructorBuilder().addModifiers(PRIVATE).build()

    private fun StorIOSQLiteColumnMeta.isNotNull(): Boolean {
        this.element.annotationMirrors.forEach {
            if (it.annotationType.toString() == ANDROID_NON_NULL_ANNOTATION) return true
        }
        return false
    }
}